### ISSUE TITLE
Allow getting projects by organization/folder

### ISSUE THREAD

**Issue by kierachell** [terraform-providers/terraform-provider-google#2958]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Currently [project data source](https://www.terraform.io/docs/providers/google/d/google_project.html) only supports getting a project by ID. But it would be nice to be able to find a project using organization/folder in case the ID is not known right away. As shown here: https://cloud.google.com/resource-manager/docs/creating-managing-folders

<img width="527" alt="screen shot 2019-01-29 at 5 07 20 pm" src="https://user-images.githubusercontent.com/18075411/51937651-61a2c780-23c8-11e9-94f8-eaf5663b0e77.png">

Seems like this should be possible since `folder_id` and `org_id` are already used in the [read function](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_google_project.go#L267)

### New or Affected Resource(s)

* google_project data source

### Potential Terraform Configuration

```tf
data "google_project" "dev" {
   organization = "org"
   folder       = "folder/path/to/it"
   name         = "dev-project"
}
```

### References

* #0000

**Comment by rileykarson** [comment]
Out of curiosity, what kind of use cases would you use this for? This is available under the https://cloud.google.com/resource-manager/reference/rest/v1/projects/list API, but searching could have unintended consequences- what if a new project with a display name that matches is added, for example?

**Comment by kierachell** [comment]
The use case is to apply similar resources within a folder: e.g. I have `main`, `dev1`, `dev2`, `dev3` projects that all might share some setup to function  (like sandbox envs).

To clarify implementation:

This would be similar to what AWS provider has for subnets: https://www.terraform.io/docs/providers/aws/d/subnet_ids.html

I want to be able to get a list of project IDs and then I can go do things with them.

**Comment by lucazz** [comment]
Heya, this is great and will fit like a glove for the module bundles I've been writing.

Ideally we would create a pre-defined GCP folder structure w/ their given projects in it and control it all using terraform's [google_project](https://www.terraform.io/docs/providers/google/r/google_project.html) and [google_folder](https://www.terraform.io/docs/providers/google/r/google_folder.html). 

That way, a person using these modules will only need to state the project name once, have everything provisioned like so and the subsequent modules like, create me a GCP VPC, create me a GKE cluster, create me a cloudSQL instance, etc could all benefit of using this datasource to fetch it's proper `project_id`.

**Comment by lucazz** [comment]
Btw, these are the modules I'm talking about:
- https://github.com/lucazz/terraform-gcp-project
- https://github.com/lucazz/terraform-gcp-network
- https://github.com/lucazz/terraform-gke
- https://github.com/lucazz/terraform-gke-node-pool

**Comment by rileykarson** [comment]
Thanks for the insight! I'm looking at implementing this so it could be configured like so:

```hcl
data "google_projects" "my-org-projects" {
  filter = "parent.id:012345678910 lifecycleState:DELETE_REQUESTED"
}

data "google_project" "deletion-candidate" {
   project_id = "${data.google_projects.my-org-projects.projects.0.project_id}"
}
```

Directly exposing the `filter` field so that we support every filter type without writing a DSL in the provider, and meaning if new filters are ever added we pick them up for free.

Does this match both of your expectations/use cases?

**Comment by lucazz** [comment]
@rileykarson what would work if we count interpolate variables in the filters. Do you know wether or not that possible?

**Comment by rileykarson** [comment]
Oh yup! Interpolations will work as you'd expect, including the dependency graph if you're using a resource in your config. For an example, using an `organization_id` variable would look like:

```hcl
data "google_projects" "my-org-projects" {
  filter = "parent.id:${var.organization_id} lifecycleState:DELETE_REQUESTED"
}

data "google_project" "deletion-candidate" {
   project_id = "${data.google_projects.my-org-projects.projects.0.project_id}"
}
```

**Comment by lucazz** [comment]
Perfect! sign me up 'cause I'm on board!

**Comment by rileykarson** [comment]
Done! This will be available in `2.2.0`.

**Comment by lucazz** [comment]
Oh man! Thats awesome!
